modules/google-analytics/wp-google-analytics.php

```diff
@@ -0,0 +1,408 @@
+	}
+
+	/**
+	 * Used to generate a tracking URL
+	 *
+	 * @param array $track - Must have ['data'] and ['code'].
+	 * @return string - Tracking URL
+	 */
+	private function _get_url( $track ) {
+		$site_url = ( is_ssl() ? 'https://':'http://' ) . sanitize_text_field( wp_unslash( $_SERVER['HTTP_HOST'] ) ); // Input var okay.
+		foreach ( $track as $k => $value ) {
+			if ( strpos( strtolower( $value ), strtolower( $site_url ) ) === 0 ) {
+				$track[ $k ] = substr( $track[ $k ], strlen( $site_url ) );
+			}
+			if ( 'data' === $k ) {
+				$track[ $k ] = preg_replace( '/^https?:\/\/|^\/+/i', '', $track[ $k ] );
+			}
+
+			// This way we don't lose search data.
+			if ( 'data' === $k && 'search' === $track['code'] ) {
+				$track[ $k ] = rawurlencode( $track[ $k ] );
+			} else {
+				$track[ $k ] = preg_replace( '/[^a-z0-9\.\/\+\?=-]+/i', '_', $track[ $k ] );
+			}
+
+			$track[ $k ] = trim( $track[ $k ], '_' );
+		}
+		$char = ( strpos( $track['data'], '?' ) === false ) ? '?' : '&amp;';
+		return str_replace( "'", "\'", "/{$track['code']}/{$track['data']}{$char}referer=" . rawurlencode( isset( $_SERVER['HTTP_REFERER'] ) ? $_SERVER['HTTP_REFERER'] : '' ) ); // Input var okay.
+	}
+
+	/**
+	 * Maybe output or return, depending on the context
+	 */
+	private function _output_or_return( $val, $maybe ) {
+		if ( $maybe ) {
+			echo $val . "\r\n";
+		} else {
+			return $val;
+		}
+	}
+
+	/**
+	 * This injects the Google Analytics code into the footer of the page.
+	 *
+	 * @param bool[optional] $output - defaults to true, false returns but does NOT echo the code.
+	 */
+	public function insert_code( $output = true ) {
+		// If $output is not a boolean false, set it to true (default).
+		$output = ( false !== $output);
+
+		$tracking_id = $this->_get_options( 'code' );
+		if ( empty( $tracking_id ) ) {
+			return $this->_output_or_return( '<!-- Your Google Analytics Plugin is missing the tracking ID -->', $output );
+		}
+
+		// get our plugin options.
+		$wga = $this->_get_options();
+		// If the user's role has wga_no_track set to true, return without inserting code.
+		if ( is_user_logged_in() ) {
+			$current_user = wp_get_current_user();
+			$role = array_shift( $current_user->roles );
+			if ( 'true' === $this->_get_options( 'ignore_role_' . $role ) ) {
+				return $this->_output_or_return( '<!-- Google Analytics Plugin is set to ignore your user role -->', $output );
+			}
+		}
+
+		// If $admin is true (we're in the admin_area), and we've been told to ignore_admin_area, return without inserting code.
+		if ( is_admin() && ( ! isset( $wga['ignore_admin_area'] ) || 'false' !== $wga['ignore_admin_area'] ) ) {
+			return $this->_output_or_return( '<!-- Your Google Analytics Plugin is set to ignore Admin area -->', $output );
+		}
+
+		$custom_vars = array(
+			"_gaq.push(['_setAccount', '{$tracking_id}']);",
+		);
+
+		// Add custom variables specified by the user.
+		foreach ( $this->_get_options( 'custom_vars', array() ) as $i => $custom_var ) {
```

Comment:
Will there be a field to set up custom vars in Calypso? If not, maybe we can remove all related code, along with the tokens?